### ISSUE TITLE
Multiple declarations for a symbol

### ISSUE THREAD

**Issue by simonbuchan** [oxc-project/oxc#11935]
Not sure if this is in scope, but given the valid typescript code:

```ts
let x: number = 42;

interface x {
    value: number;
}
```

Currently `Semantic` will be created with a single symbol for `x` (correct), but `Semnatic::symbol_declaration(id)` will return whichever `AstNode` was first in the source. This seems fairly bad, but I'm not sure how you would best approach it.

Typescript maintains a list of declarations, where it treats the first "value" declaration separately:

https://github.com/microsoft/TypeScript/blob/02672d281c26e561708127da1d8d1a6cae45fee2/src/compiler/types.ts#L6049-L6050

```ts
export interface Symbol {
    flags: SymbolFlags;                     // Symbol flags
    escapedName: __String;                  // Name of symbol
    declarations?: Declaration[];           // Declarations associated with this symbol
    valueDeclaration?: Declaration;         // First value declaration of the symbol
    ...
```

So maybe something like:

```rust
impl Semantic {
  fn symbol_value_declaration(&self, id: SymbolId) -> Option<&Declaration>;
  fn symbol_declarations(&self, id: SymbolId) -> impl Iterator<Item =&Declaration>;
}
```

Not sure what the workaround would be at the moment short of doing semantic analysis myself - doesn't seem like `Semantic` generates references for types.

**Comment by Dunqing** [comment]
We stored multiple declarations information in different field, see https://github.com/oxc-project/oxc/blob/1e88dba68bc1528a0b299ca638cece0dafab2a4b/crates/oxc_semantic/src/scoping.rs#L203-L212.

**Comment by Dunqing** [comment]
Feel free to reopen if that can't solve your problem!

**Comment by simonbuchan** [comment]
Huh, don't know how I missed something called "redeclarations" when looking for this! Sorry for the noise.

It seems like the intended use here is roughly:

```rust
for id in scoping.symbol_ids() {
  if let decls = scoping.symbol_redeclarations(id)
     && !decls.is_empty() {
    for decl in decls {
      // something with decl.*
    }
  } else {
    // something with scoping.symbol_*
  }
}
```

Is that right? It seems a bit odd, is that a memory access optimizing approach?

**Comment by Dunqing** [comment]
> Huh, don't know how I missed something called "redeclarations" when looking for this! Sorry for the noise.
> 
> It seems like the intended use here is roughly:
> 
> for id in scoping.symbol_ids() {
>   if let decls = scoping.symbol_redeclarations(id)
>      && !decls.is_empty() {
>     for decl in decls {
>       // something with decl.*
>     }
>   } else {
>     // something with scoping.symbol_*
>   }
> }
> Is that right? It seems a bit odd, 

What's your use case? 

> is that a memory access optimizing approach?

Not only that, but it also provides easier access because multiple declarations are quite rare. You rarely need to deal with that.

**Comment by simonbuchan** [comment]
> What's your use case?

Trying to see if I can port [centi-typescript](https://github.com/sandersn/mini-typescript/tree/centi-typescript) a bit easier with better machinery in place - so far yeah, pretty much! My [previous stab petered out](https://github.com/simonbuchan/tts) around unification of generics but the basics aren't actually that hard. I'm not intending to try to implement all of typescript, don't worry :P

> Not only that, but it also provides easier access because multiple declarations are quite rare. You rarely need to deal with that.

Perhaps, but it's weird to get either an empty list from `symbol_redeclarations(id)` or at least 2 items. Fine if it's documented of course.

I'm also not sure about the "You rarely need to deal with that" if you're e.g. writing a lint as you don't really get to pick the source code. A user would see issues like "this isn't used" depending on the order of declarations, which wouldn't make any sense to them. It's a stronger argument if it returned the first _value_ declaration.
